examples/dreambooth/train_dreambooth_lora_sdxl.py

```diff
@@ -506,45 +516,69 @@ def __getitem__(self, index):
-    add_text_embeds = [example["instance_added_cond_kwargs"]["text_embeds"] for example in examples]
     add_time_ids = [example["instance_added_cond_kwargs"]["time_ids"] for example in examples]
+
+    if not train_text_encoder:
+        input_ids = [example["instance_prompt_ids"] for example in examples]
+        add_text_embeds = [example["instance_added_cond_kwargs"]["text_embeds"] for example in examples]
+    else:
+        tokens_one = [example["instance_prompt_tokens_one"] for example in examples]
+        tokens_two = [example["instance_prompt_tokens_two"] for example in examples]
+
     if has_attention_mask:
         attention_mask = [example["instance_attention_mask"] for example in examples]
 
     # Concat class and instance examples for prior preservation.
     # We do this to avoid doing two forward passes.
     if with_prior_preservation:
-        input_ids += [example["class_prompt_ids"] for example in examples]
         pixel_values += [example["class_images"] for example in examples]
-        add_text_embeds += [example["class_added_cond_kwargs"]["text_embeds"] for example in examples]
         add_time_ids += [example["class_added_cond_kwargs"]["time_ids"] for example in examples]
 
+        if not train_text_encoder:
+            input_ids += [example["class_prompt_ids"] for example in examples]
+            add_text_embeds += [example["class_added_cond_kwargs"]["text_embeds"] for example in examples]
+        else:
+            tokens_one += [example["class_prompt_tokens_one"] for example in examples]
+            tokens_two += [example["class_prompt_tokens_two"] for example in examples]
+
         if has_attention_mask:
             attention_mask += [example["class_attention_mask"] for example in examples]
 
     pixel_values = torch.stack(pixel_values)
     pixel_values = pixel_values.to(memory_format=torch.contiguous_format).float()
-
-    input_ids = torch.cat(input_ids, dim=0)
-    add_text_embeds = torch.cat(add_text_embeds, dim=0)
     add_time_ids = torch.cat(add_time_ids, dim=0)
 
-    batch = {
-        "input_ids": input_ids,
-        "pixel_values": pixel_values,
-        "unet_added_conditions": {"text_embeds": add_text_embeds, "time_ids": add_time_ids},
-    }
+    if not train_text_encoder:
+        input_ids = torch.cat(input_ids, dim=0)
+        add_text_embeds = torch.cat(add_text_embeds, dim=0)
+    else:
+        tokens_one = torch.cat(tokens_one, dim=0)
+        tokens_two = torch.cat(tokens_two, dim=0)
+
+    unet_added_conditions = {"time_ids": add_time_ids}
+    if not train_text_encoder:
+        unet_added_conditions.update({"text_embeds": add_text_embeds})
+        batch = {"input_ids": input_ids, "pixel_values": pixel_values, "unet_added_conditions": unet_added_conditions}
+    else:
+        batch = {
+            "tokens_one": tokens_one,
+            "tokens_two": tokens_two,
+            "pixel_values": pixel_values,
+            "unet_added_conditions": unet_added_conditions,
+        }
```

Comment:
I believe there's actually a different refactor that could make this script (and a few of the other ones) a bit simpler.

The only data that belongs in the dataloader is the image data. All other outputs of the collate function, `input_ids`, `unet_added_conditions`, `attention_mask`, `tokens_one`, and `tokens_two` are statically determined and don't require any loading from files, etc...

If you were to just remove those components from the dataloader, we could simplify the dataloader and the PR could instead directly handle the text prompt components in the script body, thoughts?

Reply:
I like the idea, however to have self-containment in the script, I think the dataloaders are okay as is.

Reply:
Sorry but I'm not really sure I follow, could you elaborate on what this has to do with the script being self-contained? I'm not recommending making changes to the dataloaders in other scripts, just noting that the other scripts share the same issue.

I feel relatively strongly that we should not be passing already computed data to the dataloader